### PR TITLE
feat(parser): add Whispers scanner for identifying secrets

### PR DESCRIPTION
**Pull request by adeptex**
Adding [Whispers](https://github.com/adeptex/whispers) - scanner for identifying secrets in structured static text.

### CODE REVIEW

File: dojo/fixtures/test_type.json [hunk 1]
```diff
@@ -47,5 +47,12 @@
...
         "model": "dojo.test_type",
         "pk": 7
+    },
+    {
+        "fields": {
+            "name": "Whispers"
+        },
+        "model": "dojo.test_type",
+        "pk": 8
     }
```
Comment (damiencarol) [inline dojo/fixtures/test_type.json:57]: this is not needed, this file is still here for historical reasons

File: dojo/tools/whispers/parser.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                    "sensitive information in code."
+                ),
+                references=Endpoint.from_uri(
+                    "https://cwe.mitre.org/data/definitions/798.html"
+                ),
+                cwe=798,
+                severity=SEVERITY_MAP.get(vuln.get("severity"), "Info"),
+                file_path=vuln.get("file"),
+                line=int(vuln.get("line")),
+                hash_code=hashlib.sha256(str(description).encode("utf-8")).hexdigest(),
```
Comment (damiencarol) [inline dojo/tools/whispers/parser.py:64]: `hash_code` is used in another way. if you want to leverage the finger printing features of defectdojo, you should use other attributes like `uniq_id_from_tool` and/or `vuln_id_from_tool`.
this is a blocking issue

File: dojo/tools/whispers/parser.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            # internal de-duplication
+            dupe_key = finding.hash_code
+            if dupe_key in dupes:
+                find = dupes[dupe_key]
+                if finding.description:
+                    find.description += "\n" + finding.description
+                find.unsaved_endpoints.extend(finding.unsaved_endpoints)
+                dupes[dupe_key] = find
+            else:
+                dupes[dupe_key] = finding
```
Comment (damiencarol) [inline dojo/tools/whispers/parser.py:79]: internal de-duplication is not a very good design, are you sure you want to have this auto-aggregation feature in the parser?

File: dojo/tools/whispers/parser.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    def convert_severity(self, num_severity):
+        """Convert severity value"""
+        if num_severity >= -10:
+            return "Low"
+        elif -11 >= num_severity > -26:
+            return "Medium"
+        elif num_severity <= -26:
+            return "High"
+        else:
+            return "Info"
```
Comment (damiencarol) [inline dojo/tools/whispers/parser.py:92]: this method seems not used. dead code?

**Review by damiencarol — CHANGES_REQUESTED** [review]

**Comment by Maffooch** [comment]
The [how to write a parser](https://defectdojo.github.io/django-DefectDojo/contributing/how-to-write-a-parser/) guide may be of help here

**Comment by adeptex** [comment]
Hey @damiencarol @Maffooch, thanks very much for taking the time to review. All requested changes implemented.
I initially followed the guide, and created the parser using `cookiecutter https://github.com/DefectDojo/cookiecutter-scanner-parser`, which is what created all that additional code.. I wasn't sure if it was needed..

File: dojo/fixtures/test_type.json (hunk 1 above)
Comment (adeptex) [reply dojo/fixtures/test_type.json]: This was automatically added by the cookiecutter..

File: dojo/tools/whispers/parser.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                    references=Endpoint.from_uri("https://cwe.mitre.org/data/definitions/798.html"),
+                    cwe=798,
+                    severity=self.SEVERITY_MAP.get(
+                        vuln.get("severity"),
+                        "Info"),
+                    file_path=vuln.get("file"),
+                    line=int(
+                        vuln.get("line")),
+                    vuln_id_from_tool=hashlib.sha256(
+                        str(description).encode("utf-8")).hexdigest(),
```
Comment (damiencarol) [inline dojo/tools/whispers/parser.py:64]:
```suggestion
                    vuln_id_from_tool=vuln.get('key'),
```

File: dojo/settings/settings.dist.py [hunk 6]
```diff
@@ -1123,6 +1123,7 @@ def saml2_attrib_map_format(dict):
     'Hydra Scan': ['title', 'description'],
     'DrHeader JSON Importer': ['title', 'description'],
     'PWN SAST': ['title', 'description'],
+    'Whispers': ['vuln_id_from_tool'],
```
Comment (damiencarol) [inline dojo/settings/settings.dist.py:1126]:
```suggestion
    'Whispers': ['vuln_id_from_tool', 'file_path', 'line'],
```
I have the feeling that this version should be better.

File: unittests/tools/test_whispers_parser.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        parser = WhispersParser()
+        findings = parser.get_findings(testfile, Test())
+        testfile.close()
+        for finding in findings:
+            for endpoint in finding.unsaved_endpoints:
+                endpoint.clean()
+        self.assertEqual(1, len(findings))
+        self.assertEqual("src/pip.conf", findings[0].file_path)
+        self.assertEqual(2, findings[0].line)
+
```
Comment (damiencarol) [inline unittests/tools/test_whispers_parser.py:26]: could you had a test on the vuln_id_from_tool attribute?

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: unittests/tools/test_whispers_parser.py (hunk 7 above)
Comment (adeptex) [reply unittests/tools/test_whispers_parser.py:27]: added

File: dojo/settings/settings.dist.py (hunk 6 above)
Comment (adeptex) [reply dojo/settings/settings.dist.py]: added

File: dojo/tools/whispers/parser.py (hunk 5 above)
Comment (damiencarol) [reply dojo/tools/whispers/parser.py]: good

**Review by damiencarol — APPROVED** [review]

**Review by mtesauro — APPROVED** [review]
Approved

**Comment by rafaelsilvaaa10** [comment]
Hello, I'm trying to add import file and not working for me. 

https://github.com/DefectDojo/django-DefectDojo/issues/7097

**Comment by damiencarol** [comment]
> Hello, I'm trying to add import file and not working for me.
> 
> #7097

Please use issue feature of GitHub.

**Pull request merged**